File: services/elixir-phoenix/lib/stt_gateway/protocol/partial_message.ex
defmodule SttGateway.Protocol.PartialMessage do
  @moduledoc false

  @partial_type "partial"

  @derive {JSON.Encoder,
           only: [
             :type,
             :oldest_frame_seq,
             :newest_frame_seq,
             :frames,
             :rms,
             :zero_crossings,
             :checksum,
             :samples,
             :transcript,
             :audio_bytes,
             :cpu_passes,
             :model_delay_ms,
             :flush_lateness_ms,
             :inflight_model_jobs
           ]}
  defstruct [
    :type,
    :oldest_frame_seq,
    :newest_frame_seq,
    :frames,
    :rms,
    :zero_crossings,
    :checksum,
    :samples,
    :transcript,
    :audio_bytes,
    :cpu_passes,
    :model_delay_ms,
    :flush_lateness_ms,
    :inflight_model_jobs
  ]

  @type t :: %__MODULE__{
          type: String.t(),
          oldest_frame_seq: non_neg_integer(),
          newest_frame_seq: non_neg_integer(),
          frames: non_neg_integer(),
          rms: float(),
          zero_crossings: non_neg_integer(),
          checksum: non_neg_integer(),
          samples: non_neg_integer(),
          transcript: String.t(),
          audio_bytes: non_neg_integer(),
          cpu_passes: non_neg_integer(),
          model_delay_ms: non_neg_integer(),
          flush_lateness_ms: float(),
          inflight_model_jobs: non_neg_integer()
        }

  @type result :: %{
          rms: float(),
          zero_crossings: non_neg_integer(),
          checksum: non_neg_integer(),
          samples: non_neg_integer(),
          transcript: String.t(),
          audio_bytes: non_neg_integer()
        }

  @type context :: %{
          oldest_seq: non_neg_integer(),
          newest_seq: non_neg_integer(),
          frames: non_neg_integer(),
          cpu_passes: non_neg_integer(),
          model_delay_ms: non_neg_integer(),
          flush_lateness_ms: float(),
          inflight_model_jobs: non_neg_integer()
        }

  @spec new!(result(), context()) :: t()
  def new!(result, context) do
    %__MODULE__{
      type: @partial_type,
      oldest_frame_seq: context.oldest_seq,
      newest_frame_seq: context.newest_seq,
      frames: context.frames,
      rms: result.rms,
      zero_crossings: result.zero_crossings,
      checksum: result.checksum,
      samples: result.samples,
      transcript: result.transcript,
      audio_bytes: result.audio_bytes,
      cpu_passes: context.cpu_passes,
      model_delay_ms: context.model_delay_ms,
      flush_lateness_ms: context.flush_lateness_ms,
      inflight_model_jobs: context.inflight_model_jobs
    }
  end

  @spec to_json!(t()) :: binary()
  def to_json!(%__MODULE__{} = message), do: JSON.encode!(message)
end

File: services/java-helidon-nima/src/main/java/stt/protocol/PartialMessage.java
package stt.protocol;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyOrder;

/**
 * {@code partial} message sent back to the load generator on each successful flush. Field order
 * matches the Rust / Go gateways for line-by-line diff-ability.
 */
@JsonPropertyOrder({
    "type",
    "oldest_frame_seq",
    "newest_frame_seq",
    "frames",
    "rms",
    "zero_crossings",
    "checksum",
    "samples",
    "transcript",
    "audio_bytes",
    "cpu_passes",
    "model_delay_ms",
    "flush_lateness_ms",
    "inflight_model_jobs"
})
public record PartialMessage(
        String type,
        @JsonProperty("oldest_frame_seq") long oldestFrameSeq,
        @JsonProperty("newest_frame_seq") long newestFrameSeq,
        int frames,
        double rms,
        @JsonProperty("zero_crossings") long zeroCrossings,
        long checksum,
        long samples,
        String transcript,
        @JsonProperty("audio_bytes") long audioBytes,
        @JsonProperty("cpu_passes") int cpuPasses,
        @JsonProperty("model_delay_ms") long modelDelayMs,
        @JsonProperty("flush_lateness_ms") double flushLatenessMs,
        @JsonProperty("inflight_model_jobs") int inflightModelJobs)
        implements OutboundMessage {

    public static PartialMessage of(
            InferResponse infer,
            long oldestSeq,
            long newestSeq,
            int frames,
            int cpuPasses,
            long modelDelayMs,
            double flushLatenessMs) {
        return new PartialMessage(
                Protocol.PARTIAL_TYPE,
                oldestSeq,
                newestSeq,
                frames,
                infer.rms(),
                infer.zeroCrossings(),
                infer.checksum(),
                infer.samples(),
                infer.transcript(),
                infer.audioBytes(),
                cpuPasses,
                modelDelayMs,
                flushLatenessMs,
                0);
    }
}
